File: src/main/java/math/Combination.java
package math;

/**
 * nCm % MOD を求める。O(n)
 */
public class Combination {
  private long[] fact;
  private long[] invFact;
  private int MOD;

  public Combination(int max, int MOD) {
    this.MOD = MOD;
    long[] inv = new long[max + 1];
    fact = new long[max + 1];
    invFact = new long[max + 1];
    inv[1] = 1;
    for (int i = 2; i <= max; i++) inv[i] = inv[MOD % i] * (MOD - MOD / i) % MOD;
    fact[0] = invFact[0] = 1;
    for (int i = 1; i <= max; i++) fact[i] = fact[i - 1] * i % MOD;
    for (int i = 1; i <= max; i++) invFact[i] = invFact[i - 1] * inv[i] % MOD;
  }

  public long get(int x, int y) {
    return fact[x] * invFact[y] % MOD * invFact[x - y] % MOD;
  }
}

File: scala/src/main/scala/Combination.scala

/**
  * calculate nCm modulo mod
  *
  * @param max max number of n
  * @param mod modulo
  */
class Combination(max: Int, mod: Int) {
  private val inv = new Array[Long](max + 1)
  private val fact = new Array[Long](max + 1)
  private val invFact = new Array[Long](max + 1)
  inv(1) = 1
  for (i <- 2 to max) inv(i) = inv(mod % i) * (mod - mod / i) % mod
  fact(0) = 1
  invFact(0) = 1
  for (i <- 1 to max) fact(i) = (fact(i - 1) * i) % mod
  for (i <- 1 to max) invFact(i) = (invFact(i - 1) * inv(i)) % mod

  /**
    * get nCm
    */
  def get(n: Int, m: Int): Int = {
    fact(n) * invFact(m) % mod * invFact(n - m) % mod
  }.toInt
}
